src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConnectionManager.java

```diff
@@ -252,44 +252,56 @@ private void addShutdownListenerToChannel(Channel channel, Boolean withRecovery)
+
                         if (withRecovery && recoverableErrors.stream().anyMatch(errorString::contains)) {
-                            int channelNumber =  channel.getChannelNumber();
-                            var pinIdToChannelHolder = getChannelHolderByChannelNumber(channelNumber);
-                            if ((pinIdToChannelHolder != null) && (pinIdToChannelHolder.getValue().state() != ChannelState.SUBSCRIBING)) {
-                                recoverSubscriptionsOfChannel(channelNumber);
-                            }
+                            var pinId = pinIdToChannelHolder.getKey();
+                            var holder = pinIdToChannelHolder.getValue();
+                            recoverSubscriptionsOfChannel(pinId, channel, holder);
                         }
                     }
                 }
             }
         });
     }
 
-    private @Nullable Map.Entry<PinId, ChannelHolder> getChannelHolderByChannelNumber(int channelNumber) {
+    private @Nullable Map.Entry<PinId, ChannelHolder> getChannelHolderByChannel(Channel channel) {
         return channelsByPin
                 .entrySet()
                 .stream()
-                .filter(entry -> Objects.nonNull(entry.getValue().channel) && channelNumber == entry.getValue().channel.getChannelNumber())
+                .filter(entry -> channel == entry.getValue().channel)
                 .findAny()
                 .orElse(null);
     }
 
-    private void recoverSubscriptionsOfChannel(int channelNumber) {
+    private void recoverSubscriptionsOfChannel(@NotNull final PinId pinId, Channel channel, @NotNull final ChannelHolder holder) {
         channelChecker.execute(() -> {
             try {
-                var pinIdToChannelHolder = getChannelHolderByChannelNumber(channelNumber);
-                if (pinIdToChannelHolder != null) {
-                    PinId pinId = pinIdToChannelHolder.getKey();
-                    ChannelHolder channelHolder = pinIdToChannelHolder.getValue();
+                holder.lock.lock();
+                SubscriptionCallbacks subscriptionCallbacks = holder.subscriptionCallbacks;
+                boolean resubscribe = false;
+                try {
+                    if (holder.channel != channel) {
+                        LOGGER.warn("Channel already recovered");
+                        return;
+                    }
 
-                    SubscriptionCallbacks subscriptionCallbacks = channelHolder.subscriptionCallbacks;
                     if (subscriptionCallbacks != null) {
                         channelsByPin.remove(pinId);
-                        if (channelHolder.state == ChannelState.SUBSCRIBED) {
-                            LOGGER.info("Changing channel for holder with pin id: " + pinId.toString());
-                            basicConsume(pinId.queue, subscriptionCallbacks.deliverCallback, subscriptionCallbacks.cancelCallback);
+                        if (holder.isSubscribed) {
```

Comment:
Why do we operate on the internal holder state? In my understanding the lock and responsibility to track whether the channel is subscribing or not is fully on the channel's holder. But right now we manage that outside the holder. Could you please advise on why have you decided to go this way?